fuse_tutorials/src/range_sensor_simulator.cpp

```diff
@@ -117,10 +117,18 @@ std::vector<Beacon> createNoisyBeacons(const std::vector<Beacon>& beacons)
 /**
  * @brief Convert the set of beacons into a pointcloud for visualization purposes
  */
-sensor_msgs::PointCloud2::ConstPtr beaconsToPointcloud(const std::vector<Beacon>& beacons)
+sensor_msgs::PointCloud2::ConstPtr beaconsToPointcloud(
+  const std::vector<Beacon>& beacons,
+  rclcpp::node_interfaces::NodeClockInterface::SharedPtr clock_interface=nullptr)
```

Comment:
Since the `clock_interface` is used to get a temporary clock for one call, I'd recommend this method takes a `const rclcpp::Clock &` instead of the interface.

Reply:
I'll take in an `rclcpp::Clock::SharedPtr`

Reply:
https://github.com/locusrobotics/fuse/pull/283/commits/59d10ddc448f2a21928d5104dabe494e0e37f8e4

Reply:
`rclcpp::Clock::SharedPtr` as a parameter signals [that the passed in clock is going to keep the Clock alive longer than just the function call](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-sharedptrparam-owner).

I see that the intent of making clock optional is so it can fallback to a system time clock; however, this function is in a `.cpp` file [and is not exposed in the header file](https://github.com/methylDragon/fuse/blob/rolling-time/fuse_tutorials/include/fuse_tutorials/range_sensor_model.h). All calls to it [do pass in a clock](https://github.com/methylDragon/fuse/blob/ad6056c7f890ec51e65664cef7ebad02a8c3bca3/fuse_tutorials/src/range_sensor_simulator.cpp#L368-L372), so the the fallback behavior of using system time isn't needed. I think [`const rclcpp::Clock &` is a better fit](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const).

Reply:
https://github.com/locusrobotics/fuse/pull/283/commits/c18049be8cda5cf7ee3dd4eac7f3e2f1da50578e

I actually wanted to ask about this one, does this mean that the `is_valid` and `wait_for_valid` methods [in my time patch](https://github.com/locusrobotics/fuse/pull/283/files#diff-5e9d38b7ea0ad9239343157fb2e49683467e170e6df1471fce6a0d2c1a97a798R62) should also take in const refs?

I was thinking to use shared_ptrs because the node interface returns a shared ptr to the node's clock, so it'd avoid needing to do a dereference, and also because, in my view, the node's clock would continue to exist.

Reply:
> I actually wanted to ask about this one, does this mean that the is_valid and wait_for_valid methods [in my time patch](https://github.com/locusrobotics/fuse/pull/283/files#diff-5e9d38b7ea0ad9239343157fb2e49683467e170e6df1471fce6a0d2c1a97a798R62) should also take in const refs?

Probably, yeah :no_mouth: 

> I was thinking to use shared_ptrs because the node interface returns a shared ptr to the node's clock, so it'd avoid needing to do a dereference, and also because, in my view, the node's clock would continue to exist.

Fair point, a `Clock::SharedPtr` is more convenient given other APIs return that.

> and also because, in my view, the node's clock would continue to exist.

From the perspective of `beaconsToPointcloud()`, the clock only has to exist while the it's being run.